Get iOS VPN widget ready to ship

graeme opened the pull request:
Task/Issue URL: https://app.asana.com/0/0/1207255718764347/f

**Description**:

- Get ready to ship the iOS home screen widget, as specced in [Figma designs](https://www.figma.com/file/tTKJAnjILeUEUhxCBeFyhI/Hackdays-Widget?type=design&node-id=84-7954&mode=design&t=Rmgi8wPqcEvRvp1N-0)
- Update the VPN settings view to include instructions for adding the widget

**Steps to test this PR**:
[Steps in this task](https://app.asana.com/0/0/1207255817447243/f)

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

graeme committed:
Update VPN widget gallery text

graeme committed:
Update VPN widget connection status messages

graeme committed:
Use disconnected text when vpn status is invalid

graeme committed:
Add period to gallery description text

graeme committed:
Update disconnected subtitle

graeme committed:
Localise widget button text

graeme committed:
Update widget images

graeme committed:
Tweak colors and layout to match designs

graeme committed:
Use disconnected state for placeholder

graeme committed:
Show disconnected state on error / notConfigured

graeme committed:
Update gallery text to mention Privacy Pro

graeme committed:
Add VPN Widget education screen

graeme committed:
Fix settings item styling

graeme committed:
Tweaks to education view for new images

graeme committed:
Update widget education iphone image

graeme committed:
Link to app when VPN is not configured or error

graeme committed:
Direct user to privacy pro when linking to VPN

graeme committed:
Set placeholder widget to connected to LA

graeme committed:
Final copy updates

graeme committed:
Update widget education view's accent colour

graeme committed:
Update location widget label to only include city

graeme commented on Core/UserDefaults+NetworkProtection.swift:
```diff
@@ -33,7 +33,7 @@ public extension UserDefaults {
 public enum NetworkProtectionUserDefaultKeys {
 
-    public static let lastSelectedServer = "com.duckduckgo.network-protection.last-selected-server"
+    public static let lastSelectedServerCity = "com.duckduckgo.network-protection.last-selected-server-city"
```
This is now only the city so want to make sure it’s not used for anything expecting more (currently only used for the widget).

graeme commented on DuckDuckGo/AppDelegate.swift:
```diff
@@ -858,10 +858,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
-                if visibility.shouldShowVPNShortcut() {
-                    presentNetworkProtectionStatusSettingsModal()
-                }
+                presentNetworkProtectionStatusSettingsModal()
```
Now always handling the unsubscribed case in the function itself.

graeme commented on DuckDuckGo/NetworkProtectionVPNSettingsView.swift:
```diff
@@ -127,4 +136,18 @@ struct NetworkProtectionVPNSettingsView: View {
+            widgetExampleImageConfig: .init(
+                image: Image("WidgetEducationVPNWidgetExample"),
+                maxWidth: 164,
+                horizontalOffset: -7
```
This is gross but I couldn’t think of another way (it’s because the assets have shadow which differ…)

graeme commented on DuckDuckGo/WidgetEducationView.swift:
```diff
@@ -27,8 +27,33 @@ extension Font {
+    }
+}
+
 struct WidgetEducationView: View {
```
Changes in this file are to allow for the VPN specific widget education

graeme commented on DuckDuckGo/WidgetEducationView.swift:
```diff
@@ -101,7 +130,7 @@ private struct NumberedCircle: View {
     static let background = Color(designSystemColor: .background)
     static let font = Color("WidgetEducationFontColor")
-    static let circle = Color(UIColor.cornflowerBlue)
+    static let circle = Color(designSystemColor: .accent)
```
Squeezed in a general styling update here while it’s being changed on Robert’s request.

graeme commented on Widgets/VPNWidget.swift:
```diff
@@ -50,18 +51,18 @@ class VPNStatusTimelineProvider: TimelineProvider {
         NETunnelProviderManager.loadAllFromPreferences { managers, error in
             let defaults = UserDefaults.networkProtectionGroupDefaults
-            let location = defaults.string(forKey: NetworkProtectionUserDefaultKeys.lastSelectedServer) ?? "Unknown Location"
+            let location = defaults.string(forKey: NetworkProtectionUserDefaultKeys.lastSelectedServerCity) ?? "Unknown Location"
```
The country didn’t always fit so we decided to use only the city for the widget.

graeme committed:
Make the VPN intents non-discoverable

graeme committed:
Enable VPN widget for non-alpha builds

quanganhdo commented on DuckDuckGo/en.lproj/Localizable.strings:
```diff
@@ -116,7 +116,7 @@
 "addWidget.description" = "Get quick access to private search and the sites you love.";
 
 /* No comment provided by engineer. */
```
Nit: Missing comment.

quanganhdo commented on DuckDuckGo/en.lproj/Localizable.strings:
```diff
@@ -2303,6 +2303,9 @@ But if you *do* want a peek under the hood, you can find more information about
 /* Alert title for the alert when the Privacy Pro subscription expires */
 "vpn.access-revoked.alert.title" = "VPN disconnected due to expired subscription";
 
+/* No comment provided by engineer. */
```
Also missing comment.

graeme committed:
VPN Widget Pixels (#2832)

graeme committed:
Remove unused color

graeme committed:
Reload widget on VPN connection status change

graeme committed:
Update on/off copy

graeme committed:
Update darkmode education asset

graeme committed:
Various styling tweaks after ship review feedback

samsymons committed:
Merge branch 'main' into graeme/ship-ios-vpn-widget

* main: (27 commits)
  Release 7.120.0-0 (#2846)
  Enable gzip compression for Sync PATCH payloads (#2805)
  sync error notify (#2822)
  Wire up VPN error UI (#2826)
  Mergeback from release/7.119.0
  Release 7.119.0-1 (#2840)
  Add temporary survey to Passwords screen (#2834)
  Upload logs on cancel (#2836)
  Add bookmarks to top hits suggestions on iOS (#2835)
  Fix Maestro reporting for e2e tests (#2713)
  support marketplace kit (#2833)
  Enable attribution flag (#2825)
  Release 7.119.0-0 (#2831)
  Fix subscriptions glitch in Settings (#2828)
  Delegate SecureVault KeyStore Events in All Scenarios  (#2827)
  Clears subscription cache on restore (#2823)
  Enable field validation for Sync payloads (#2807)
  Fix up address bar spoofing tests to match new behaviours. (#2743)
  Release 7.118.0-4 (#2821)
  Settings: onAppear changed to onDisappear (#2817) (#2820)
  ...

samsymons committed:
Update VPN string comment.

samsymons committed:
Update widget background colors (#2847)

Task/Issue URL: https://app.asana.com/0/414235014887631/1207310384147176/f
Tech Design URL:
CC:

Description:

This PR updates each widget to use the correct background color, as requested by the design and product teams.

samsymons committed:
Merge branch 'main' into graeme/ship-ios-vpn-widget

* main: (42 commits)
  Remove run loop use, use DI for Bookmarks Caching Search (#2854)
  Bump BSK Version (#2843)
  Release 7.120.0-3 (#2861)
  Update BSK that reverts broken site pixel encoding change (#2860)
  Tweak sync Maestro tests (#2858)
  Release 7.120.0-2 (#2856)
  Fix for how we encode broken site reports (#2855)
  New Autofill pixels for KPIs (#2849)
  Add support for `targetPercentile` RMF rule parameter (#2824)
  Release 7.120.0-1 (#2853)
  Uncomment test code
  Use different function to check if in portrait
  use old provisioning profile template for the alpha build (#2848)
  Address Ship Review requests
  Update tests to handle new Settings UI (#2844)
  Fix more swiftlint issues
  Fix swiftlint issues
  Fix brainfart
  Don't show the toast when there are still some DaxDialogs left
  Make sure we fire reload, on just going back from background
  ...

samsymons committed:
Fix a navigation pathing animation issue.

samsymons committed:
Reuse the existing settings VC if necessary.

samsymons committed:
Allow the widgets to read and write VPN configurations.

samsymons commented on Widgets/WidgetsExtension.entitlements:
```diff
@@ -2,6 +2,10 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+	<key>com.apple.developer.networking.networkextension</key>
+	<array>
+		<string>packet-tunnel-provider</string>
+	</array>
```
Targets need to have the VPN entitlement in order to read/write changes. I'll be honest: I have no idea how this was working before this. But, myself and Jackson was able to reproduce occasional `permission denied` errors, and fixing the entitlements looks to have solved it completely. 🤷

samsymons commented on Widgets/WidgetViews.swift:
```diff
@@ -168,7 +168,7 @@ struct FavoritesWidgetView: View {
     var body: some View {
         ZStack {
-            Rectangle().fill(Color(designSystemColor: .surface))
+            Rectangle().fill(Color(designSystemColor: .backgroundSheets))
```
These changes were reviewed in another PR stacked on top of this one, and requested by the product team.

samsymons commented on DuckDuckGo/MainViewController+Segues.swift:
```diff
@@ -256,14 +256,21 @@ extension MainViewController {
-        present(navController, animated: true) {
-            completion?(settingsViewModel)
+
+        if let navigationController = self.presentedViewController as? UINavigationController,
+           let settingsHostingController = navigationController.viewControllers.first as? SettingsHostingController {
+            navigationController.popToRootViewController(animated: false)
+            completion?(settingsHostingController.viewModel)
+        } else {
+            let settingsController = SettingsHostingController(viewModel: settingsViewModel, viewProvider: legacyViewProvider)
+
+            // We are still presenting legacy views, so use a Navcontroller
+            let navController = UINavigationController(rootViewController: settingsController)
+            settingsController.modalPresentationStyle = UIModalPresentationStyle.automatic
+
+            present(navController, animated: true) {
+                completion?(settingsViewModel)
+            }
         }
```
Please give these changes a thorough review - the issue this solves is the case where you already have the settings UI open and you tap the Connect button without being authenticated. Before this change, you wouldn't get sent to the correct child view controller. Now, we reuse the existing settings controller and push the change onto it.
